infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java

```diff
@@ -0,0 +1,180 @@
+                );
+    }
+
+    /**
+     * Reads data entries by utilizing the replication protocol. If there are errors, retry.
+     * If the wrong epoch exception occurs, fail immediately, so that we can retry the workflow
+     * earlier.
+     *
+     * @param transferBatchRequest The transferBatchRequest of addresses to be read.
+     * @return A result of reading records.
+     */
+    @VisibleForTesting
+    CompletableFuture<ReadBatch> readRecords(TransferBatchRequest transferBatchRequest) {
+        return CompletableFuture.supplyAsync(() -> {
+
+            Optional<ReadBatchException> latestReadException = Optional.empty();
+            Optional<ReadBatch> latestReadBatch = Optional.empty();
+            for (int i = 0; i < maxReadRetries; i++) {
+                try {
+                    Map<Long, ILogData> records =
+                            addressSpaceView.simpleProtocolRead(transferBatchRequest.getAddresses(), readOptions);
+                    ReadBatch batch = checkReadRecords(transferBatchRequest.getAddresses(),
+                            records, transferBatchRequest.getDestination());
+                    latestReadBatch = Optional.of(batch);
+                    if (batch.getStatus() == ReadBatch.ReadStatus.FAILED) {
+                        throw new IllegalStateException("Some addresses failed to transfer: " +
+                                batch.getFailedAddresses());
+                    } else {
+                        latestReadException = Optional.empty();
+                        break;
+                    }
+
+                } catch (WrongEpochException e) {
+                    log.warn("readRecords: encountered a wrong epoch exception on try {}: ", i, e);
+                    latestReadException = Optional.of(new ReadBatchException(e));
+                    break;
+                } catch (RuntimeException e) {
+                    log.warn("readRecords: encountered an exception on try {}: ", i, e);
+                    latestReadException = Optional.of(new ReadBatchException(e));
+                    Sleep.sleepUninterruptibly(readSleepDuration);
+                }
+            }
+
+            if (latestReadBatch.isPresent() && latestReadException.isPresent()) {
+                // Batch was read, but the retries were exhausted because of an exception.
+                return latestReadBatch.get().toBuilder().causeOfFailure(latestReadException).build();
+            } else if (latestReadBatch.isPresent()) {
+                // No latest read failure, return a read batch.
+                return latestReadBatch.get();
+            } else {
+                // Batch not present, only failure.
+                return ReadBatch.builder()
+                        .failedAddresses(transferBatchRequest.getAddresses())
+                        .status(ReadBatch.ReadStatus.FAILED)
+                        .causeOfFailure(latestReadException)
+                        .build();
```

Comment:
I feel there is a problem here. The result (a ReadBatch) of the readRecord() is directly fed to writeRecords(), however the  writeRecords() does not check the state of the ReadBatch, it directly uses ReadBatch.getData() to write. If the readRecords() is a failure, writeRecords() will not know and just try to write an empty list of data because ReadBatch.getData() returns empty? Should we return an exceptionally completed future in this case?

Reply:
I decided to handle that case in the handle block after the read, so that we can fail the future earlier. I have also written a test.

Reply:
I approved but the logic here and the new error handling logic seems a bit twisted. I think you can simply throw exceptions here whenever latestReadException is present instead of a bunch of if else blocks and an extra error handling block after read which introduces more if elses.